### ISSUE TITLE
Relax `Matrix` to `AbstractMatrix` etc.

### ISSUE THREAD

**Issue by JeffFessler** [JuliaMath/NFFT.jl#74]
As I try to update to 0.8 I am hitting issues here because I pass in a transposed `Matrix` so it is not a `Matrix` type but rather an `AbstractMatrix` type.  Can we relax it?  (Same with `Array` to `AbstractArray`, just in case.)

https://github.com/JuliaMath/NFFT.jl/blob/7a13b3afc1f025642cdbdcc1a606906bd2690668/src/NFFT.jl#L42

I can provide a PR if it seems reasonable...

**Comment by tknopp** [comment]
This should actually work, see 

    https://github.com/JuliaMath/NFFT.jl/blob/master/AbstractNFFTs/src/derived.jl#L19

can you give me a 3-liner that I can execute on the REPL, then I can investigate the issue.

**Comment by tknopp** [comment]
Ah, I think the issue is the first argument. That will be very difficult to allow a transpose there. The first argument is used to dispatch between `Array` and `CuArray`. But both are `AbstractArrays`.

**Comment by JeffFessler** [comment]
This 1-liner does it:

```julia
plan_nfft(rand(3,2)', (8,10))

ERROR: MethodError: no method matching plan_nfft(::Type{Array}, ::LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, ::Tuple{Int64, Int64})
Closest candidates are:
  plan_nfft(::Type{<:Array}, ::Matrix{T}, ::Tuple{Vararg{Int64, D}}, ::Any...; timing, kargs...) where {T, D} at ~/.julia/packages/NFFT/mIegb/src/NFFT.jl:43
  plan_nfft(::Type, ::AbstractVector, ::Tuple{Vararg{Int64, D}}, ::Any...; kwargs...) where D at ~/.julia/packages/AbstractNFFTs/zd135/src/derived.jl:16
  plan_nfft(::AbstractArray, ::Union{Integer, Tuple{Vararg{Int64, D}}}, ::Any...; kargs...) where D at ~/.julia/packages/AbstractNFFTs/zd135/src/derived.jl:8
```

But I just realized I am doing another convert elsewhere and I can simply transpose at that point since I am allocating anyway, and anyway this is just a 1-time thing for the plan, so it is fine with me to close this issue if you want.

**Comment by tknopp** [comment]
Thanks, thats a bug, please leave this open.

**Comment by tknopp** [comment]
should work now

**Comment by JeffFessler** [comment]
Thanks!  I'll take you word for it.  I tried to verify myself by using `NFFT#master` but got an error at `using NFFT` like this:

```
[ Info: Precompiling NFFT [efe261a4-0d2b-5849-be55-fc731d526b0d]
ERROR: LoadError: UndefVarError: AbstractNNFFTPlan not defined
Stacktrace:
  [1] top-level scope
    @ ~/.julia/packages/NFFT/hcCCN/src/direct.jl:11
  [2] include(mod::Module, _path::String)
    @ Base ./Base.jl:418
  [3] include(x::String)
    @ NFFT ~/.julia/packages/NFFT/hcCCN/src/NFFT.jl:1
  [4] top-level scope
    @ ~/.julia/packages/NFFT/hcCCN/src/NFFT.jl:30
  [5] include
    @ ./Base.jl:418 [inlined]
  [6] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
    @ Base ./loading.jl:1318
  [7] top-level scope
    @ none:1
  [8] eval
    @ ./boot.jl:373 [inlined]
  [9] eval(x::Expr)
    @ Base.MainInclude ./client.jl:453
 [10] top-level scope
    @ none:1
in expression starting at /Users/fessler/.julia/packages/NFFT/hcCCN/src/direct.jl:11
in expression starting at /Users/fessler/.julia/packages/NFFT/hcCCN/src/NFFT.jl:1
```
So I'll just wait for the next release - it's fine.  Impressive turn around there :)

**Comment by tknopp** [comment]
yes, master is unstable so I realized that my fix will not help you. But I triggered the release. Need to make this package by package (AbstractNFFTs -> NFFT -> NFFTTools). This release will be  v0.9 for NFFT and it will have some interface changes. As indicated before I wanted to change `nfft!` to `mul!` and this is what has been done now. Furthermore triggered by the contact to FINUFFT devs we now also have the type 3 NFFT as an abstract type.
